analyzer/runtime/evm/client.go

```diff
@@ -149,90 +156,48 @@ func evmCallWithABI(
-		// later) only if the error is non-deterministic, e.g. network
-		// failure. Otherwise, accept that the token contract doesn't expose
-		// this info.
-		if !errors.Is(err, EVMDeterministicError{}) {
-			return nil, fmt.Errorf("calling name: %w", err)
-		}
-	}
-	if err := evmCallWithABI(ctx, source, round, tokenEthAddr, evmabi.ERC20, &tokenData.Symbol, "symbol"); err != nil {
-		logError("symbol", err)
-		if !errors.Is(err, EVMDeterministicError{}) {
-			return nil, fmt.Errorf("calling symbol: %w", err)
-		}
-	}
-	if err := evmCallWithABI(ctx, source, round, tokenEthAddr, evmabi.ERC20, &tokenData.Decimals, "decimals"); err != nil {
-		logError("decimals", err)
-		if !errors.Is(err, EVMDeterministicError{}) {
-			return nil, fmt.Errorf("calling decimals: %w", err)
-		}
-	}
-	mutable, err := evmDownloadTokenERC20Mutable(ctx, logger, source, round, tokenEthAddr)
-	if err != nil {
-		return nil, err
-	}
-	if mutable == nil {
-		return nil, nil
-	}
-	tokenData.EVMTokenMutableData = mutable
-	return &tokenData, nil
+// logDeterministicError is for if we know how to handle a deterministic
+// error--in those cases you can use this to make a note of the error. Just in
+// case someone wasn't expecting it, you know?
+func logDeterministicError(logger *log.Logger, round uint64, contractEthAddr []byte, interfaceName string, method string, err error, keyvals ...interface{}) {
+	keyvals = append([]interface{}{
+		"round", round,
+		"contract_eth_addr_hex", hex.EncodeToString(contractEthAddr),
+		"interface_name", interfaceName,
+		"method", method,
+		"err", err,
+	}, keyvals...)
+	logger.Info("call failed", keyvals...)
 }
 
 // EVMDownloadNewToken tries to download the data of a given token. If it
 // transiently fails to download the data, it returns with a non-nil error. If
 // it deterministically cannot download the data, it returns a struct
 // with the `Type` field set to `EVMTokenTypeUnsupported`.
 func EVMDownloadNewToken(ctx context.Context, logger *log.Logger, source nodeapi.RuntimeApiLite, round uint64, tokenEthAddr []byte) (*EVMTokenData, error) {
-	// todo: check ERC-165 0xffffffff compliance
-	// todo: try other token standards based on ERC-165
-	// see https://github.com/oasisprotocol/oasis-indexer/issues/225
+	useERC165, err := detectERC165(ctx, logger, source, round, tokenEthAddr)
+	if err != nil {
+		return nil, fmt.Errorf("detect ERC165: %w", err)
+	}
+	if useERC165 {
+		supportsERC721, ok, err1 := detectInterface(ctx, logger, source, round, tokenEthAddr, ERC721InterfaceID)
+		if err1 != nil {
+			return nil, fmt.Errorf("checking ERC721 interface: %w", err1)
+		}
+		if ok && supportsERC721 {
+			tokenData, err2 := evmDownloadTokenERC721(ctx, logger, source, round, tokenEthAddr)
+			if err2 != nil {
+				return nil, fmt.Errorf("download token ERC-721: %w", err2)
+			}
+			return tokenData, nil
+		}
+
+		// todo: add support for other token types
+		// see https://github.com/oasisprotocol/oasis-indexer/issues/225
+	}
 
 	// Check ERC-20.
```

Comment:
```suggestion
	// Check ERC-20. These tokens typically do not implement ERC-165, so do not bother checking for
	// explicit declaration of support, detect ERC-20 support with heuristics instead.
```

Reply:
ok

Reply:
or rather I'll have it say that ERC-20 doesn't require ERC-165 support